plugin/src/main/java/com/playmonumenta/scriptedquests/managers/SongManager.java

```diff
@@ -207,11 +218,59 @@ public static boolean stopSong(Player player, boolean cancelNow) {
 		return state.nextSong();
 	}
 
+	public static void playSongWhile(Collection<Player> players, Song song, boolean playNow, BooleanSupplier shouldPlay, boolean cancelNow, int cancelDelay, int checkInterval) {
+		Collection<Player> songPlayers = new ArrayList<>(players);
+		if (shouldPlay.getAsBoolean()) {
+			playSong(songPlayers, song, playNow);
+
+			new BukkitRunnable() {
+				@Override
+				public void run() {
+					// Remove players whose song was canceled by something else
+					songPlayers.removeIf(p -> {
+						Song current = getCurrentSong(p);
+						return current == null || !song.equalsSong(current);
+					});
+
+					if (songPlayers.isEmpty()) {
+						this.cancel();
+						return;
+					}
+
+					if (!shouldPlay.getAsBoolean()) {
+						Bukkit.getScheduler().runTaskLater(Plugin.getInstance(), () -> stopSong(songPlayers, cancelNow), cancelDelay);
+						this.cancel();
+					}
+				}
+			}.runTaskTimer(Plugin.getInstance(), checkInterval, checkInterval);
+		}
+	}
+
+	public static void playSongWhile(Player player, Song song, boolean playNow, BooleanSupplier shouldPlay, boolean cancelNow, int cancelDelay, int checkInterval) {
+		playSongWhile(List.of(player), song, playNow, shouldPlay, cancelNow, cancelDelay, checkInterval);
+	}
+
+	public static void playBossSong(Collection<Player> players, Song song, boolean playNow, LivingEntity boss, boolean cancelNow, int cancelDelay, int checkInterval) {
+		playSongWhile(players, song, playNow, () -> boss.isValid() && !boss.isDead(), cancelNow, cancelDelay, checkInterval);
+	}
+
+	public static void playBossSong(Player player, Song song, boolean playNow, LivingEntity boss, boolean cancelNow, int cancelDelay, int checkInterval) {
+		playBossSong(List.of(player), song, playNow, boss, cancelNow, cancelDelay, checkInterval);
+	}
+
 	public static void onLogout(Player player) {
 		PlayerState state = mPlayerStates.remove(player.getUniqueId());
 		if (state != null) {
 			state.cancelNext();
 			state.cancelNow();
 		}
 	}
+
+	public static void onDeath(Player player) {
+		PlayerState state = mPlayerStates.remove(player.getUniqueId());
+		if (state != null && state.mNow != null && state.mNow.mStopOnDeath) {
```

Comment:
If `state.mNow.mStopOnDeath` is false, this loses track of the current state of the player.